group-edit.ejs 실제 데이터 반영하여 표현하기

youngToMaturity opened the pull request:
`group-edit` 페이지는 `mypage`에서 사용자가 특정 스터디 그룹의 "관리" 버튼을 클릭할 시 해당 스터디 그룹의 정보를 보여주고, 수정할 수 있는 페이지 입니다.
따라서 이 브랜치에서 구현하고자 하는 것은 아래와 같습니다.
* `study-group`의 정보에 맞게 접속시 default 값이 정해져있는 것 
>현재는 `대분류`,`중분류`,`온/오프라인`, `성별`, `인원수`, `스터디 이름`으로 되어있는 페이지의 기본값을
`취업`, `미디어`, `온라인`, `남성만`, `3인`, `카카오 가자`와 같이 채우는 것

해당 내용을 구현하기 전, 당시의 `develop` 브랜치의 `mypage`의 관리 버튼은 특정 스터디의 `study_group_id`를 담고 있지 않았기 때문에, 임시로 해당 정보를 버튼에 담아주기 위한 과정이 필요했습니다.

### '관리' 버튼 클릭시 `study_group_id`가 `group-edit` 페이지에 전달되도록 하는 과정
 가장 먼저, `public/js/mypage.js` 파일에서 `window`가 `load`되면 `fetch`가 실행되도록 하였고,
`routes/mypage.js`에서 `req.user.id`와 다중 쿼리를 통해 `req.user.id`를 `id`로 갖는 `group_member` 테이블들을 `is_manager` 값에 따라 두개의 배열로 나누어 돌려주었습니다.(이 부분이 정확히 이해되지 않습니다 // result와 data의 관계, 어떤 형태로 전달되는 것인지)

이후, result를 data로 받아온 후, `study_group_id`를 value로 가질 `input`을 `type='hidden'`으로 각 스터디에 생성해주고,  `forEach` 함수를 통해 `data[0][count].study_group_id`를 value로 지정하였습니다. 변수 `count`는 초기값이 0이고 forEach가 실행될때마다 1씩 증가합니다.

마지막으로 '관리' 버튼을 클릭시 `form.submit()`을 통해 `hidden`타입의 `input`이 갖고있는 `value`, `study_group_id`를 `group-edit/process`에 post 하도록 했습니다.

### `group-edit/process`에서 `study_group_id`를 통해 스터디 정보와 스터디원 전달하기
`routes/mypage.js`에서 `req.body.study_group_id`를 통해 다중 쿼리문을 작성하였습니다. 해당 쿼리의 결과로, 전달받은 `study_group_id`와 같은 `id`를 갖고있는 `study_group` 데이터와 `study_group_id`가 같은 `group_members` 데이터를 `group-edit`을 render 할 때 같이 전달했습니다.

youngToMaturity committed:
feat: mypage 로드시, 로그인 된 user가 가입/생성한 그룹의 데이터를 fetch

각 스터디 클릭시 group-edit에 group_id를 넘길 수 있도록 임시로 실행하고자 하는 과정. (mypage가 완성되면 수정 예정)

youngToMaturity committed:
feat: fetch를 통해 받은 study_gorup_id를 각 group-list__edit-btn의 value로 지정

[임시]
! issue : 각 lead-group-list의 group-list__item의 갯수를 어떻게 지정할 것인가? - user마다 운영하고 있는 스터디 그룹의 갯수가 다를것이기 때문에
! issue: group-list__item에 해당 study_group_id까지는 지정할 수 있겠지만, 해당 div의 내용들을 어떻게 바꿀 것인가? (fetch속의 fetch?)

youngToMaturity committed:
feat : 관리 버튼 클릭시 각 item에 해당하는 study-group-id post

관리 버튼 클릭시 group-list-hidden__input에 저장된 study-group-id가 post된 후 group-edit이 render되도록 함

youngToMaturity committed:
feat: '관리' 버튼 클릭시 post로 group-edit.ejs에 선택된 그룹의 정보 넘기기

db.query를 활용하여 submit된 id와 같은 study_group을 찾고, 해당 그룹에 속해있는 group_member도 result로 저장. 하지만 ejs에 대한 이해도 부족으로 이 값들을 ejs 상에 표시하는것 미완성

42KIM committed:
feat: Load lead-group list by ssr+csr

42KIM commented:
study_group_id를 전달하는 과정에서요!

제가 마이페이지 렌더링 파트 구현하면서, 마이페이지에 접속 시
진행 중인 스터디 리스트를 렌더링하면서 영수님이 만들어놓으신 hidden input에 해당 study_group_id를 value로 넣어주는 과정이 포함되어 있습니다. 
(영수님과 마찬가지로 fetch를 통해 클라이언트 사이드에서 넣어주는 방법, 서버사이드 렌더링으로 넣어주는 방법 둘 다 있습니다)
렌더링 과정에서 이미 필수적으로 수행되는 작업이니 이 부분 감안해서 진행하셔도 될 것 같아요!!
아 영수님이 직접 구현해보시는 것도 당연히 좋다고 생각합니다! 혹시나 해당 부분이 포함되어 있는 걸 발견 못하실까 해서 말씀드렸습니다ㅎㅎ

youngToMaturity committed:
style :  ejs 적용과정 공유를 위한 커밋

youngToMaturity committed:
feat : group-edit 페이지에서 기존의 스터디 그룹 정보를 기본값으로 표시해주기

수정하기 페이지에 들어가는 경우, 해당 스터디의 기존 정보를 page에 보이도록 하였습니다. ( 중분류, 멤버 미구현 )

youngToMaturity commented:
## 추가 사항 1
### `group-edit.ejs` 파일에서 전달받은 데이터를 기반으로 기본 값 세팅하기
앞선 `group-edit/process`에서 전달받은 `name`,`location`,`gender`,`members`,`main`,`sub`,`member`의 데이터를 수정하기 페이지에 들어왔을때 기본 값으로 세팅해주도록 하였습니다. 각 `<option>` 태그를 반복문을 통해 생성하도록 하였고, 그 `value`와 전달받은 데이터가 같은 경우 `<option>`태그에 `selected` 옵션이 추가되도록 하여 구현하였습니다.

youngToMaturity commented:
> study_group_id를 전달하는 과정에서요!
> 
> 제가 마이페이지 렌더링 파트 구현하면서, 마이페이지에 접속 시
> 진행 중인 스터디 리스트를 렌더링하면서 영수님이 만들어놓으신 hidden input에 해당 study_group_id를 value로 넣어주는 과정이 포함되어 있습니다.
> (영수님과 마찬가지로 fetch를 통해 클라이언트 사이드에서 넣어주는 방법, 서버사이드 렌더링으로 넣어주는 방법 둘 다 있습니다)
> 렌더링 과정에서 이미 필수적으로 수행되는 작업이니 이 부분 감안해서 진행하셔도 될 것 같아요!!
> 아 영수님이 직접 구현해보시는 것도 당연히 좋다고 생각합니다! 혹시나 해당 부분이 포함되어 있는 걸 발견 못하실까 해서 말씀드렸습니다ㅎㅎ

네! 처음 브랜치 생성할때 구현된 부분이 없었어서 임시로 만들어서 진행했습니다! merge시에 조율해서 진행하면 될 것 같습니다!

42KIM commented on routes/mypage.js:
```diff
@@ -26,6 +26,20 @@ router.get('/', (req, res, next) => {
   }
 });
 
+router.get('/fetch', (req, res, next) => {
```
마지막에 수정할 때 조금 더 구체적인 경로명으로 수정해주면 좋을 것 같습니다. fetch 사용되는 부분이 다른 곳도 있을 테니까요!

youngToMaturity replied:
`/fetch-group`으로 수정하겠습니다!

42KIM commented on routes/mypage.js:
```diff
@@ -26,6 +26,20 @@ router.get('/', (req, res, next) => {
+  const sql_member =
+    'SELECT * FROM group_member WHERE user_id = ? AND is_manager = 0';
+  db.query(sql_manager + sql_member, [id, id], (err, result) => {
+    if (err) {
```
>routes/mypage.js에서 req.user.id와 다중 쿼리를 통해 req.user.id를 id로 갖는 group_member 테이블들을 is_manager 값에 따라 두개의 배열로 나누어 돌려주었습니다.

이 부분이 코드에서는 아직 두 개의 배열로 나누는 부분이 구현이 안 된 거죠?

youngToMaturity replied:
result[0]과 result[1]로 구분된 상태입니다!

youngToMaturity committed:
feat: group-edit 페이지 상의 중분류 select를 수정하고자 하는 스터디 그룹의 기존 값으로 지정

route/mypage.js의 /fetch를 /fetch-group으로 수정
js/group-edit.js의 categoryChange(e)에 있던 중분류 list를 전역으로 선언하여 DOMContentLoad시와 change시 모두 활용할 수 있도록 함.
categotyChange(e) 함수내부의 중분류 Element들의 생성과 지정을 setSub()함수로 처리하였고, hidden Input을 통해 확인할 수 있는 render시에 받은 sub과 같은 경우 selected 옵션을 추가.

42KIM committed:
refactor: Change MySQL BETWEEN to LIMIT

42KIM committed:
style: Change variable name

youngToMaturity committed:
feat: 멤버의 수만큼 div를 생성하여 닉네임과 강퇴버튼을 부여

강퇴 버튼의 value에 group-member의 id를 지정했고, 닉네임은 각 회원의 닉네임으로 지정했다. 현재는 임시로 방장도 출력했지만 추후에는 방장을 제외하고 출력하도록 처리할 것

youngToMaturity commented:
## 추가 사항 2

### 중분류 값 반영
render시에 받은 대분류와 중분류 값 `main`과 `sub`을 활용해서 `group-edit` 페이지에 접속시 대분류는 `main`이 `selected`가 되도록 했습니다. 하지만, 각각의 대분류에 따라 `options`의 내용과 개수가 달라지는 중분류의 특성에 의해, 중분류의 값은 `hidden` 속성을 가진 `input`태그에 담아서 보내고, `document`에 `DOMContentLoaded` 이벤트가 발생하는 경우, 대분류의 selected 값을 확인하여 해당 대분류의 중분류에 해당하는 list를 `option`으로 생성하고, 만약 그 값이 `input` 태그의 값과 같다면 `selected` 속성이 부여되도록 하였습니다

### 스터디원 반영
스터디원 또한 render시에 `member`로 선언된 array입니다. `ejs` 파일 내부에 `member.length`만큼 반복하는 반복문을 통해 각 멤버의 수에 맞게 `div`를 생성해주고, 해당 `div`의 내용을 `memeber[i].nickname` `member[i].id`와 같이 i번째 스터디원의 정보로 채워주도록 했습니다. 현재는 스터디장의 정보도 불러오지만, 추후에는 `is_manager`를 통해 방장은 제외한 스터디원만 반영하도록 수정이 필요합니다.

> 앞으로 구현할 사항
>* 수정된 내용이 제출되기 전 유효한 입력인지 확인하기
>* 수정된 내용들을 기반으로 수정완료 버튼 클릭시 db의 내용도 수정되도록 하기

42KIM committed:
feat: 마이페이지 다중쿼리, 서브쿼리 추가

42KIM committed:
feat: 무한스크롤 기능 구현, 더보기 버튼 삭제

youngToMaturity committed:
feat : 수정완료 버튼 클릭시 db 수정, 예외처리 두가지 적용

만약 db상에 수정하고자 하는 이름이 있고 && 그 이름이 현재 수정하고자 하는 스터디 그룹의 이름이 아니라면 수정불가.
만약 현재 스터디 인원수보다 수정하고자 하는 스터디 정원이 더 적다면 수정불가.
그 외의 상황은 수정 가능

youngToMaturity committed:
feat: 강퇴 버튼 클릭시 해당 멤버 강퇴 기능 구현

group-edit.ejs - form 위치 수정 (css 재지정 필요)
routes/mypage.js - kickoutForm post 처리, 방장인 경우(추후 방장은 members에 보이지 않도록 처리 요망) 강퇴안됨
js/mypage.js - Kickout시 방생되는 status 2개 추가, 기존의 complete은 edit으로 수정 ?: 각 case 처리 이후에 redirect할 수 있을까?
js/group-edit.js 강퇴 버튼 클릭시 kickout 폼 submit

youngToMaturity commented:
## 추가 사항 3
### 수정완료시 db의 내용도 수정
수정완료 버튼을 누르는 경우 `form`이 `submit`되어 수정내용이 반영되도록 했습니다. 정상적인 작동을 위해 예외처리도 지정해주었는데, 처리한 예외는 다음의 두가지였습니다.
* `수정하고자 하는 스터디 그룹의 이름이 이미 존재하는 경우` && `해당 이름을 가진 스터디 그룹이 현재 바꾸고자 하는 그룹이 아닌 경우`
이 두가지 조건을 충족시키는 경우 이미 다른 스터디 그룹이 해당 이름을 사용하는 것이기 때문에 db를 수정하지 않고, `mypage`로 돌아가면서 `?status=invalidname`을 추가해주어 이미 사용중인 이름임을 `alert` 하도록 했습니다.
* `수정하고자 하는 정원보다 현재 스터디 그룹 참여자의 인원이 더 큰 경우`
수정하고자 하는 정원보다 현재 인원이 더 많은 경우 스터디 그룹이 정원보다 많은 참여 인원을 갖게되기 때문에, db를 수정하지 않고 `?status=invalidmember`를 통해 해당 정보를 사용자에게 `alert`해주었습니다.

그 외의 경우에는 db를 사용자가 입력한 값으로 수정해주고, `?status=complete`(다음 커밋에서 edit으로 수정)으로 정상적으로 처리됨을 사용자에게 알렸습니다.

### 멤버의 강퇴 기능 구현
강퇴 버튼에 담겨있는 `id`를 통해 해당 `user_id`를 갖고있는 `group_member` 테이블을 찾아 그 테이블을 삭제하도록 했습니다. 현재는 참여중인 멤버 목록에 스터디 그룹장도 있기 때문에 해당 예외처리를 진행했습니다.

youngToMaturity committed:
Merge remote-tracking branch 'origin/ssr+csr' into group-edit

# Conflicts:
#	public/js/mypage.js
#	routes/mypage.js
#	views/includes/mypage/lead-group-list.ejs
